File: pylib-v1/ailets/io/memkv.py
from typing import Literal, Dict, Sequence
from ailets.atyping import IKVBuffer, IKVBuffers


class MemKVBuffer(IKVBuffer):
    def __init__(self, path: str, shared_buffer: bytearray):
        self.buffer = shared_buffer
        self._path = path

    def borrow_mut_buffer(self) -> bytearray:
        return self.buffer


class MemKV(IKVBuffers):
    def __init__(self) -> None:
        self._buffers: Dict[str, bytearray] = {}

    def open(self, path: str, mode: Literal["read", "write", "append"]) -> IKVBuffer:
        if mode == "read":
            if path not in self._buffers:
                raise KeyError(f"Path not found: {path}")
            return MemKVBuffer(path, self._buffers[path])
        if mode == "write":
            buffer = bytearray()
            self._buffers[path] = buffer
            return MemKVBuffer(path, buffer)
        if mode == "append":
            buffer = self._buffers.get(path, bytearray())
            self._buffers[path] = buffer
            return MemKVBuffer(path, buffer)
        raise ValueError(f"Invalid mode: {mode}")

    def flush(self, kvbuffer: IKVBuffer) -> None:
        pass

    def listdir(self, dir_name: str) -> Sequence[str]:
        if dir_name and not dir_name.endswith("/"):
            dir_name = dir_name + "/"
        return [path for path in self._buffers.keys() if path.startswith(dir_name)]

File: ailets-rs/ailetos/src/io/memkv.rs
//! In-memory implementation of KVBuffers

use super::types::{KVBuffer, KVBuffers, KVError, OpenMode};
use parking_lot::Mutex;
use std::collections::HashMap;
use std::sync::Arc;

/// In-memory implementation of KVBuffers
///
/// Simple hash map based storage, useful for testing and single-process use.
pub struct MemKV {
    buffers: Mutex<HashMap<String, KVBuffer>>,
}

impl MemKV {
    /// Create a new empty MemKV store
    #[must_use]
    pub fn new() -> Self {
        Self {
            buffers: Mutex::new(HashMap::new()),
        }
    }
}

impl Default for MemKV {
    fn default() -> Self {
        Self::new()
    }
}

impl KVBuffers for MemKV {
    async fn open(&self, path: &str, mode: OpenMode) -> Result<KVBuffer, KVError> {
        let mut buffers = self.buffers.lock();

        match mode {
            OpenMode::Read => buffers
                .get(path)
                .cloned()
                .ok_or_else(|| KVError::NotFound(path.to_string())),
            OpenMode::Write => {
                let buffer = Arc::new(Mutex::new(Vec::new()));
                buffers.insert(path.to_string(), Arc::clone(&buffer));
                Ok(buffer)
            }
            OpenMode::Append => {
                if let Some(buffer) = buffers.get(path) {
                    Ok(Arc::clone(buffer))
                } else {
                    let buffer = Arc::new(Mutex::new(Vec::new()));
                    buffers.insert(path.to_string(), Arc::clone(&buffer));
                    Ok(buffer)
                }
            }
        }
    }

    async fn flush(&self, _path: &str) -> Result<(), KVError> {
        // No-op for in-memory storage
        Ok(())
    }

    async fn listdir(&self, dir_name: &str) -> Result<Vec<String>, KVError> {
        let prefix = if dir_name.ends_with('/') {
            dir_name.to_string()
        } else {
            format!("{dir_name}/")
        };

        let buffers = self.buffers.lock();
        let mut paths: Vec<String> = buffers
            .keys()
            .filter(|path| path.starts_with(&prefix))
            .cloned()
            .collect();
        paths.sort();
        Ok(paths)
    }

    async fn destroy(&self) -> Result<(), KVError> {
        let mut buffers = self.buffers.lock();
        buffers.clear();
        Ok(())
    }
}
